weavertest/deployer.go

```diff
@@ -0,0 +1,376 @@
+		GroupId:       uuid.New().String(),
+		Id:            uuid.New().String(),
+		SameProcess:   d.wlet.SameProcess,
+		Sections:      d.wlet.Sections,
+		SingleProcess: d.wlet.SingleProcess,
+		SingleMachine: d.wlet.SingleMachine,
+	}
+	conn, err := conn.NewEnvelopeConn(fromWeaveletReader, toWeaveletWriter, d, wlet)
+	if err != nil {
+		d.t.Fatalf("cannot create envelope conn: %v", err)
+	}
+	go func() {
+		// TODO(mwhittaker): Close this conn when the unit test ends. Right
+		// now, the conn lives forever. This means the pipes are also leaking.
+		// We might have to add a Close method to EnvelopeConn.
+		if err := conn.Run(); err != nil {
+			d.t.Error(err)
+		}
+	}()
+
+	bootstrap := runtime.Bootstrap{
+		ToWeaveletFile: toWeaveletReader,
+		ToEnvelopeFile: fromWeaveletWriter,
+		TestConfig:     config,
+	}
+	ctx := context.WithValue(d.ctx, runtime.BootstrapKey{}, bootstrap)
+	return weaver.Init(ctx)
+}
+
+// Stop shuts down the deployment. It blocks until the deployment is fully
+// destroyed.
+func (d *deployer) Stop() error {
+	d.mu.Lock()
+	defer d.mu.Unlock()
+
+	// Stop envelopes.
+	var errs []error
+	for _, group := range d.groups {
+		for _, envelope := range group.envelopes {
+			if err := envelope.Stop(); err != nil {
+				errs = append(errs, err)
+			}
+		}
+	}
+
+	// Wait for all goroutines to end.
+	d.stopped.Wait()
+
+	if len(errs) > 0 {
+		return errlist{errs}
+	}
+	return nil
+}
+
+// RecvLogEntry implements the envelope.EnvelopeHandler interface.
+func (d *deployer) RecvLogEntry(entry *protos.LogEntry) {
+	d.logMu.Lock()
+	defer d.logMu.Unlock()
+	if d.log {
+		// NOTE(mwhittaker): We intentionally create a new pretty printer for
+		// every log entry. If we used a single pretty printer, it would
+		// perform dimming, but when the dimmed output is interspersed with
+		// various other test logs, it is confusing.
+		d.t.Log(logging.NewPrettyPrinter(colors.Enabled()).Format(entry))
+	}
+}
+
+// RecvTraceSpans implements the envelope.EnvelopeHandler interface.
+func (d *deployer) RecvTraceSpans([]trace.ReadOnlySpan) error {
+	// Ignore traces.
+	return nil
+}
+
+// ReportLoad implements the envelope.EnvelopeHandler interface.
+func (d *deployer) ReportLoad(*protos.WeaveletLoadReport) error {
+	// Ignore load.
+	return nil
+}
+
+// GetAddress implements the envelope.EnvelopeHandler interface.
+func (d *deployer) GetAddress(req *protos.GetAddressRequest) (*protos.GetAddressReply, error) {
+	return &protos.GetAddressReply{Address: "localhost:0"}, nil
+}
+
+// ExportListener implements the envelope.EnvelopeHandler interface.
+func (d *deployer) ExportListener(req *protos.ExportListenerRequest) (*protos.ExportListenerReply, error) {
+	return &protos.ExportListenerReply{}, nil
+}
+
+// RegisterReplica implements the envelope.EnvelopeHandler interface.
+func (d *deployer) RegisterReplica(req *protos.ReplicaToRegister) error {
+	d.mu.Lock()
+	defer d.mu.Unlock()
+
+	group := d.group(req.Group)
+	routing, _ := group.routing.Get("")
```

Comment:
This conflicts with my versioned.Versioned change, but I can merge it on top.

Reply:
We chatted offline, and I'm rebasing on top of the `versioned.Versioned` change.